### PR TITLE
Revert "Revert "[RLlib] Speedup A3C up to 3x (new training_iteration function instead of execution_plan) and re-instate Pong learning test.""

### PR DESCRIPTION
**Pull request by sven1977**
Reverts ray-project/ray#22250

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/examples/bare_metal_policy_with_custom_view_reqs.py [hunk 1]
```diff
@@ -35,6 +36,8 @@ def get_cli_args():
 
     config = {
         "env": "CartPole-v0",
+        # Use GPUs iff `RLLIB_NUM_GPUS` env var set to > 0.
+        "num_gpus": int(os.environ.get("RLLIB_NUM_GPUS", "0")),
```
Comment (gjoliver) [inline rllib/examples/bare_metal_policy_with_custom_view_reqs.py:40]: just confirming, is this the only example that is missing num_gpus config?
all the other tests we updated here have this bit already right?
just don't want to waste that gpu if they get scheduled this way.

**Review by gjoliver — APPROVED** [review]
man this is exciting!
how do I tell whether a test was actually run on gpu or not??

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

File: rllib/examples/bare_metal_policy_with_custom_view_reqs.py (hunk 1 above)
Comment (sven1977) [reply rllib/examples/bare_metal_policy_with_custom_view_reqs.py]: GPU don't really make sense for A3C, as all gradient computations (the most expensive part of the algo) happen on the workers.

Comment (sven1977) [reply rllib/examples/bare_metal_policy_with_custom_view_reqs.py]: I don't think we use the `RLLIB_NUM_GPUS` setting a lot. Only in the CI in 1-2 tests (example scripts).

**Pull request merged**